src/Swift.Runtime/src/Swift/Runtime/SwiftHandle.cs

```diff
@@ -90,58 +41,71 @@ public static implicit operator IntPtr(SwiftHandle value)
     /// <summary>
     /// Explicit conversion from IntPtr to SwiftHandle
     /// </summary>
-    public static explicit operator SwiftHandle(IntPtr value)
+    public static implicit operator SwiftHandle(IntPtr value)
     {
         return new SwiftHandle(value);
     }
+}
 
+/// <summary>
+/// Represents an opaque handle to a Swift object of type T.
+/// Used to manage native memory associated with a Swift object of type T.
+/// </summary>
+public sealed class SwiftSafeHandle<T> : SafeHandleZeroOrMinusOneIsInvalid where T : ISwiftObject
+{
     /// <summary>
-    /// Explicit conversion from SwiftHandle to void*
+    /// Returns a SwiftSafeHandle with a zero value
     /// </summary>
-    public unsafe static explicit operator void*(SwiftHandle value)
-    {
-        return (void*)(IntPtr)value;
-    }
+    public readonly static SwiftSafeHandle<T> Zero = new SwiftSafeHandle<T>(IntPtr.Zero);
 
     /// <summary>
-    /// Explicit conversion from void* to SwiftHandle
+    /// Constructs a SwiftSafeHandle from the given IntPtr
     /// </summary>
-    public unsafe static explicit operator SwiftHandle(void* value)
+    public SwiftSafeHandle(IntPtr handle)
+        : base(ownsHandle: true)
     {
-        return new SwiftHandle((IntPtr)value);
+        SetHandle(handle);
     }
 
     /// <summary>
-    /// Compare this SwiftHandle to the given object, returns true if they're the same.
+    /// Releases the handle to the Swift object
     /// </summary>
-    public override bool Equals(object? o)
+    protected override unsafe bool ReleaseHandle()
     {
-        if (o is SwiftHandle nh)
-            return nh.handle == this.handle;
-        return false;
-    }
+        TypeMetadata metadata = SwiftObjectHelper<T>.GetTypeMetadata();
+        metadata.ValueWitnessTable->Destroy((void*)handle, metadata);
 
-    /// <summary>
-    /// Generate a hashcode for the SwiftHandle
-    /// </summary>
-    public override int GetHashCode()
-    {
-        return handle.GetHashCode();
+        NativeMemory.Free((void*)handle);
+        handle = IntPtr.Zero;
+
+        return true;
     }
+}
 
-    /// <summary>
-    /// Returns true if this matches the give SwiftHandle
-    /// </summary>
-    public bool Equals(SwiftHandle other)
+/// <summary>
+/// Represents a buffer for a Swift object used for lowering.
+/// </summary>
+public sealed class PayloadBuffer : IDisposable
```

Comment:
Can you try making this a `ref struct` and returning it as `PayloadBuffer` instead? You won't be able to return `IDisposable` due to boxing issues with `ref struct`, but `using` supports `ref struct` types that contain a `Dispose()` method. This avoids unnecessary allocations for interop eventhough the contract is a bit looser. I think the tradeoff is worth it, as long as the current usage is okay with it.

/cc @rolfbjarne

Reply:
Good idea, updated.